### ISSUE TITLE
Receive MySQL error 'access denied for user' every time container is started

### ISSUE THREAD

**Issue by nloding** [laradock/laradock#456]
I'm using the default Laradock set up - the only thing I've changed is probably the cause of this. Every time I run `docker-compose stop` then `docker-compose up -d nginx mysql`, I cannot log in to the database. I can manually grant permissions to the homestead user, and then things work until I stop the containers again.

Here's the thing I changed: the database name is not `homestead` however, it's something else due to how I migrated the data into the container. I didn't change the database name in the docker-compose.yml file when I first set things up. I've changed it now, but have the same result.

Do I need to run `docker-compose stop` then `up` again? Will all my data be lost if I do that?

**Comment by philtrep** [comment]
@nloding There's a [script](https://github.com/laradock/laradock/blob/master/mysql/startup) that runs to create the database every time the `mysql` image is restarted. You _should_ not be losing your data by re-upping the container.

**That being said**
My suggestion (assuming you really don't want to lose your data) is that you do a dump of your existing data then move the `.sql` somewhere on your host, recreate the containers then reimport your data in the new database.

Data persistence is still pretty tricky with docker containers...

Let me know if this solves your issue!

**Comment by nloding** [comment]
@philtrep not sure I follow - if it's creating the database, why is all my data there when I do a normal `docker-compose stop` followed by `docker-compose up -d nginx mysql`? It sounds like my assumption above (changing the config after the initial container build) is the likely culprit, and that I should just destroy and rebuild the containers.

**Comment by philtrep** [comment]
@nloding The mysql files are persistent since they are saved on a shared volume (on the host). There is no "delete database" command that runs when starting the service, only a `CREATE DB IF NOT EXIST`. However, there have been times where a change in the mysql image created a loss of data when re-running the service. That is why I suggested doing a dump before the `down & up`  unless you don't care about your data.

I hope this explains things more clearly.

**Comment by nloding** [comment]
@philtrep I think it does. I'm still a bit confused on the difference between `stop` and `down`, but `down` is what I need to do to fix the issue.

**Comment by nloding** [comment]
@philtrep this didn't resolve my issue. I've done a few different things.

1. Edited docker-compose.yml and set MYSQL_DATABASE to the correct database name
2. Edited mysql/Dockerfile and set `ARG MYSQL_DATABASE` to the correct database name

I've run `docker-compose down` then `docker-compose up` after each, and I still get the permissions errors. Am I missing something, or is this truly a bug?

**Comment by philtrep** [comment]
@nloding I actually had the same problem yesterday running the latest version of LaraDock (albeit a _slightly_ modified version) in production.

The steps I took were:

1. Down the services/networks/volumes `docker-compose down`
2. Deleted all the containers `docker rm -f $(docker ps -qa)`
3. Deleted all the images `docker rmi $(docker images -qa --digests)`
4. Deleted the mysql volume `docker volume rm laradock_mysql`
5. Changed the second line in the `mysql/startup` file to `CREATE USER IF NOT EXISTS 'MYSQL_USER'@'%';`
6. Build and run the services `docker-compose up --build -d nginx mysql`

This is all from memory so I might be slightly off, I was in a rush and didn't think about writing down everything I did as I did it :rage4: 

Let me know if this works for you...

**Comment by nloding** [comment]
Doing a bit of digging, it looks like the issue is that the `mysqld --initialize` option only executes if the data directory doesn't exist or is empty. After I dumped data into the database, it's no longer empty. Here was my fix, though it seems like there should be a way to check permissions when you run `docker-compose up` ... but being new, I don't know how to do that (yet).

1. `docker-compose stop`
2. `docker-compose up --build -d mysql`
3. Tested I could log in with my user
4. `docker-compose stop`
5. `docker-compose up -d nginx mysql`
6. Tested I could still log in with my user

No data was lost during this.
